design/sketch.webidl

```diff
@@ -0,0 +1,532 @@
+};
+
+// AttachmentState
+dictionary WebGPUAttachmentStateDescriptor {
+    sequence<WebGPUTextureFormatEnum> formats;
+    // TODO other stuff like sample count etc.
+};
+
+interface WebGPUAttachmentState {
+};
+
+// Common stuff for ComputePipeline and RenderPipeline
+typedef u32 WebGPUShaderStageEnum;
+interface WebGPUShaderStage {
+    const u32 VERTEX = 0;
+    const u32 FRAGMENT = 1;
+    const u32 COMPUTE = 2;
+};
+
+dictionary WebGPUPipelineStageDescriptor {
+    WebGPUShaderModule module;
+    WebGPUShaderStageEnum stage;
+    DOMString entryPoint;
+    // TODO other stuff like specialization constants?
+};
+
+dictionary WebGPUPipelineDescriptorBase {
+    WebGPUPipelineLayout layout;
+    sequence<WebGPUPipelineStageDescriptor> stages;
+};
+
+// ComputePipeline
+dictionary WebGPUComputePipelineDescriptor : WebGPUPipelineDescriptorBase{
+};
+
+interface WebGPUComputePipeline {
+};
+
+// WebGPURenderPipeline
+typedef u32 WebGPUPrimitiveTopologyEnum;
+interface WebGPUPrimitiveTopology {
+    const u32 POINT_LIST = 0;
+    const u32 LINE_LIST = 1;
+    const u32 LINE_STRIP = 2;
+    const u32 TRIANGLE_LIST = 3;
+    const u32 TRIANGLE_STRIP = 4;
+};
+
+dictionary WebGPURenderPipelineDescriptor : WebGPUPipelineDescriptorBase{
+    WebGPUPrimitiveTopologyEnum primitiveTopology;
+    sequence<WebGPUBlendState> blendState;
+    WebGPUDepthStencilState depthStencilState;
+    WebGPUInputState inputState;
+    WebGPUAttachmentState attachmentState;
+    // TODO other properties
+};
+
+interface WebGPURenderPipeline {
+};
+// ****************************************************************************
+// COMMAND RECORDING (Command buffer and all relevant structures)
+// ****************************************************************************
+
+typedef u32 WebGPULoadOpEnum;
+interface WebGPULoadOp {
+    const u32 CLEAR = 0;
+    const u32 LOAD = 1;
+};
+
+typedef u32 WebGPUStoreOpEnum;
+interface WebGPUStoreOp {
+    const u32 STORE = 0;
+};
+
+dictionary WebGPURenderPassAttachmentDescriptor {
+    WebGPUTextureView attachment;
+    WebGPULoadOp loadOp;
+    WebGPUStoreOp storeOp;
+}
+
+dictionary WebGPURenderPassDescriptor {
+    sequence<WebGPURenderPassAttachmentDescriptor> colorAttachments;
+    WebGPURenderPassAttachmentDescriptor depthStencilAttachmaent;
+};
+
+interface WebGPUCommandBuffer {
+};
+
+interface WebGPUCommandEncoder {
+    WebGPUCommandBuffer finishEncoding();
+
+    // Commands allowed outside of "passes"
+    void copyBufferToBuffer(WebGPUBuffer src,
+                            u32 srcOffset,
+                            WebGPUBuffer dst,
+                            u32 dstOffset,
+                            u32 size);
+    // TODO figure out all the arguments required for these
+    void copyBufferToTexture();
+    void copyTextureToBuffer();
+    void copyTextureToTexture();
+    void blit();
+
+    // Allowed in both compute and render passes
+    void setPushConstants(WebGPUShaderStageFlags stage,
```

Comment:
For simplicity there's a small number like 16*4 bytes allowed on every stage separately. The D3D12 code just needs to update two spots in the root signature instead of just one.